### DISCUSSION TITLE
BUG: Wix 4 WixUI is inaccessible

### DISCUSSION THREAD

**Discussion by BryanCrotaz** [wixtoolset/issues#7834] [Questions]
Following the instructions at https://wixtoolset.org/docs/v3/wixui/wixui_dialog_library/ as there are no docs in v4

From another page it said I had to install WixToolset.UI.wixext so I've done that from NuGet

```xml
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs"
  xmlns:ui="http://wixtoolset.org/schemas/v4/wxs/ui">
  <Package 
	  Name="Sure Agent"
	  Version="1.0.0.0"
	  UpgradeCode="68c96489-4459-4f0b-8c73-e4f6f8f8ddf1"
  >
    <MajorUpgrade DowngradeErrorMessage="!(loc.DowngradeError)" />

	<Media Id="1" EmbedCab="yes" Cabinet="Files"/>
	  
    <Feature Id="Main">
      <ComponentGroupRef Id="AgentServiceFiles" />
	  <ComponentGroupRef Id="RuntimeFiles" />
	</Feature>

	  <UIRef
        Id="WixUI_Minimal"
        />
  </Package>
</Wix>

```

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	WIX0094	The identifier 'WixUI:WixUI_Minimal' is inaccessible due to its protection level.	AgentSetupWix4	H:\Projects\Client2\Client\AgentSetupWix4\Package.wxs	18	
```

**Comment by chrpai** [comment]
WiXUI now uses the WixUI element instead of the UIRef element. Since you've declared the namespace with the prefix ui it would be `<ui:WiXUI Id="WixUI_Minimal" />

I do admit the error message is a wee bit sketch.  

PS- Great ways to learn WiX v4 is to write something in v3 then use WiXCop/HeatWave to update it to v4  and then compare the differences.

**Comment by BryanCrotaz** [comment]
> WiXUI now uses the WixUI element instead of the UIRef element

Please point me to the Wix4 documentation that says how to use it.

> Great ways to learn WiX v4 is to write something in v3 then use WiXCop/HeatWave to update it to v4 and then compare the differences

Really?? I want to be productive. I want to build an installer for my new product. I don't want to learn to use an obsolete tool, just to find out how to use a different one.

**Reply by chrpai** [reply]
I'm not on the WiX team. I am just a user like you.  I do have an awesome open source project that makes authoring WiX easier and I think you'd like it. I am willing to spend an hour on a screenshare with you showing the benefits to you.

**Reply by BryanCrotaz** [reply]
send me a link and I'll take a look, but I'm now at the end of a very painful process and not that keen to repeat it. Unless it makes custom actions easy? That's the bit I've completely failed to make work without finding a workaround.

**Reply by chrpai** [reply]
Checkout https://github.com/iswix-llc/iswix-tutorials

**Reply by chrpai** [reply]
@BryanCrotaz custom actions take indepth knowledge to implement safely.  Start a thread over in discussions and I can try to help you.

**Comment by chrpai** [comment]
BTW the docs are at https://wixtoolset.org/docs/schema/ui/wixui/

WixUI element (Ui extension)
Attributes
Id (String, required) : Identifier for the dialog set to reference.
InstallDirectory (String) : An optional identifier that specifies the directory to set WIXUI_INSTALLDIR to, for WixUI dialog sets that require it.

There is also a link to submit PRs to improve the documentation.  For this page it points to 

https://github.com/wixtoolset/web/blob/master/src/xsd4/ui.xsd

I see value in this page linking to a list of the available experiences and discussions on how to extend the UI.  There is a general v4 needs moredocumentation issue so they are aware of it.

**Reply by BryanCrotaz** [reply]
> There is also a link to submit PRs to improve the documentation.

I've submitted several bug reports and PRs and have received a generally dismissive response, so I've given up on that.

Plus it's very hard to submit documentation improvements when you don't understand how to us it yourself. I even tried submitting a PR with a TODO in it hoping that someone who knew what they were doing would fill in the missing info. Nope, closed immediately with "this is not how we develop documentation" - nothing about the right way to do it, just "go away".

**Reply by BryanCrotaz** [reply]
> https://wixtoolset.org/docs/schema/ui/wixui/

I'll give you that that is indeed documentation of the available settings. But at no point does it say why you need one, or how to use it. There's no hyper link to what a dialog set is, or how to build a UI from dialogs (my god - is that complicated!). It took the entire team joined on a call all day today just to understand how an example from online worked. And even then we were stumped on how you would know how to write conditions - where is the syntax or the available flags (like Installed) documented?

It's taken 21 man days for us to build an installer that installs a single feature into a fixed install directory. That's crazy. It should have taken a couple of days at most if it was understandable from the docs. 

We have given up on a custom action to register it with the cloud server, after spending 3 days getting it to even compile, it doesn't run. Instead we're writing a registry string and having the application register itself. Not ideal as now the secret tenant key is written to disk for future hackers to find. Note that the documentation for how to write to the registry is only there in v3 - not a sausage in v4, which is a big problem because the syntax is changed bigly. Once again hours to work out how to do something that would have taken minutes if the documentation was there.

**Reply by BryanCrotaz** [reply]
PS I know this is not your fault! I've seen soooo many discussions on Stack Overflow and elsewhere that include comments about how dire the docs are dating back 15 years in some cases.

**Reply by chrpai** [reply]
Both FireGiant and myself offer consulting services. I'd love to help you.

**Reply by robmen** [reply]
> I've submitted several bug reports and PRs and have received a generally dismissive response, so I've given up on that.

That is not an accurate representation of the response to your PRs. You've submitted 4 PRs:

1. The first was incorporated with additional information from another PR.

2. In the second, you inserted a TODO comment into the website that said someone else should provide an answer. As you noted, that's not helpful or productive. If _you'd_ like to learn how to write that particular piece of documentation, then consider asking about the feature while offering to update the documentation. Someone might be more keen to explain the details you are missing, knowing that you will translate the final results into documentation for others.

3. There was a misunderstanding initially around the third PR where the GitHub rendering made it look like a redundant addition to existing documentation. Once that mistake on my part was cleared up, the PR was accepted. 

4. The fourth PR is still open with review feedback that has not been integrated yet.

The response to your PRs has not been dismissive.

On the other hand, making announcements that someone needs to add/improve documentation is highly likely to be ignored. So, that behavior of yours probably feels pretty well dismissed.

Remember, this is an open source community. You're not a customer. You will not be successful in the community by making demands or disparaging the volunteer efforts of others.

If you want to be a customer, Christopher already pointed out options for that.

**Reply by wmanning** [reply]
Likely too late to be use for you but maybe helps someone else.

There are a couple of books on WiX 3 available.  The first one I got when I first started and it helped me.  Almost required reading IMHO.

- [WiX: A Developer's Guide to Windows Installer XML](https://www.packtpub.com/product/wix-36-a-developers-guide-to-windows-installer-xml/9781782160427).  It covers MSI and burn.
-  [WiX Cookbook](https://www.packtpub.com/product/wix-cookbook/9781784393212) by the same author. 

Based on Wix 3.6 but 3.6 is very close to 3.11.  

Like @chrpai said its probably best to develop in Wix 3.x and then convert to WiX 4.x to learn how 4.x works since it's so new.  There's a lot more information out there on Wix 3 than Wix 4.  

Heat/harvesting and custom actions are advanced topics which I stayed away from but I was lucky that I was developing the software and installer at the same time so could ensure the installer was simple and smallish number of files (< 100 meant I could just add as a go).  Did develop a vary simple native custom action (c++) latter on.
